### PR TITLE
Fix (*artifact.Config).Unpack not preserving not overridden values

### LINKED ISSUE #3914: 5k+ download attempts within few seconds in agent logs while upgrading 8.11.3 agent to 8.12.0 with invalid agent binary.
**Kibana Build details:**
```
VERSION: 8.12.0 BC1
BUILD: 69840
COMMIT: 93cff0aacd70bd0835cd244348f742ba3cafc4aa
```

**Host OS:**  SLES15

**Preconditions:**
1. 8.12.0 BC1 Kibana cloud environment should be available.
2. 8.11.3 Agent should be installed using agent policy.
3. Invalid agent binary must be added.

**Steps to reproduce:**
1. Trigger upgrade for 8.11.3 agent with invalid agent binary.
2. Navigate to agent logs tab.
3. Observe 5k+ download attempts within few seconds.

**NOTE:**
- Issue is consistently reproducible every time for 8.11.3.
- We haven't observed this issue with 8.11.1 linux agent.

**Screenshot:**
image

**Expected Result:**
Download attempts shouldn't be so frequent while upgrading 8.11.3 agent to 8.12.0 with invalid agent binary.

**Logs:**
8.11.3[Issue]:
[elastic-agent-diagnostics-2023-12-14T08-19-39Z-00.zip](https://github.com/elastic/elastic-agent/files/13671367/elastic-agent-diagnostics-2023-12-14T08-19-39Z-00.zip)

8.11.1[Not an issue]:
[elastic-agent-diagnostics-2023-12-14T07-57-23Z-00.zip](https://github.com/elastic/elastic-agent/files/13671374/elastic-agent-diagnostics-2023-12-14T07-57-23Z-00.zip)

### PR DESCRIPTION
**Pull request by AndersonQ**
## What does this PR do?

(*artifact.Config).Unpack was creating a new instance of Config and manually coping all values from the original to the new one. New values added to config, RetrySleepInitDuration in this case, were neither maintaining its default value nor being updated with the new value. Now it uses reflection to ensure all values are correctly copied.

## Why is it important?

When applying a new `agent.download` config some of the default values were set to the zero value and even if they'd come on the new fleet config, they'd be ignored.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- ~~[ ] I have added an integration test or an E2E test~~

## How to test this PR locally
Try any of the following:
 - Copy the test `TestConfig_Unpack` to main and see it failing.
 - Set the `retry_sleep_init_duration` in the agent's policy, enrol it to fleet and check it's not set.
 - Reproduce the issue described in #3914

## Related issues

- Closes #3914

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

**Comment by cmacknz** [comment]
> *artifact.Config).Unpack was creating a new instance of Config and manually coping all values from the original to the new one. 

Do we know why it was doing this? Are there any hints in the commit history? This seems intentional so I am wondering if this was because of a bug or something from the past.

**Comment by cmacknz** [comment]
Ah the answer is the comment, thanks!

**Review by blakerouse — APPROVED** [review]
This is just nasty. Really sucks that this needs to be done this way. I think the way you handled it is the best way to go currently.

Really glad you added a test to ensure those structures stay inline.

**Comment by cmacknz** [comment]
Since we need to special case the transport, which includes the proxy configuration, I wonder if we should implement https://github.com/elastic/elastic-agent/issues/3820 as part of this PR to ensure we didn't break it. 

It's already in the current sprint and assigned to Anderson anyway.

**Comment by AndersonQ** [comment]
buildkite test this

**Comment by pierrehilbert** [comment]
` invalid character 'E' looking for beginning of value`

Could it be something coming from your change?

**Review by belimawr — CHANGES_REQUESTED** [review]

**Comment by jlind23** [comment]
@AndersonQ will you also tackle https://github.com/elastic/elastic-agent/issues/3820 as suggested by Craig abovE?

**Review by belimawr — APPROVED** [review]

**Comment by AndersonQ** [comment]
> @AndersonQ will you also tackle https://github.com/elastic/elastic-agent/issues/3820 as suggested by Craig abovE?

Not in this PR

File: internal/pkg/agent/application/upgrade/artifact/config.go [hunk 1]
```diff
@@ -206,20 +239,30 @@ func (c *Config) Arch() string {
...
+	// of the field in c and set in tmp.
+	// There is a test to ensure Config will always be a superset of
+	// configWithoutHTTPTransportSettings.
+	for i := 0; i < tmpType.NumField(); i++ {
+		name := tmpType.Field(i).Name
+		tmpField := tmpValue.FieldByName(name)
+
+		if tmpField.IsValid() && tmpField.CanSet() {
+			tmpField.Set(cValue.Field(i))
+		}
```
[the remark this answers is not in the document]
Comment (cmacknz) [reply internal/pkg/agent/application/upgrade/artifact/config.go]: Thanks for catching this, this suggests there is a test case missing to catch this.

**Comment by AndersonQ** [comment]
buildkite test this

**Pull request merged**